packages/std/src/deps.rs

```diff
@@ -22,3 +22,43 @@ impl<S: Storage, A: Api, Q: Querier> Deps<S, A, Q> {
         }
     }
 }
+
+pub struct DepsMut<'a, S: Storage, A: Api, Q: Querier> {
+    pub storage: &'a mut S,
```

Comment:
Not sure I totally understand the problem (or the full space of the problem) but, if you want to provide discretionary mutable access to `Storage`, why not wrap `Storage` itself into a `RefCell`, i. e.:

```
pub storage: RefCell<&'a S>
```

That way, you can continue to use `Deps` as before, and just `borrow_mut()` storage when mutable access to storage is required.

Reply:
I believe I would need RefCell<S> to be able to get a &mut S from &RefCell.

The cell can't give out something it never had

Reply:
One reason motivating this (before removing generics) was that previously the caller had to **own** S and Q to be able to give a reference out. If the owner has &mut S and &Q that is all the access the handle function needs, but it could not actually call it. (I hit this trying to build a test framework).

Once I started using references, the idea of removing generics and using &dyn came up and got rid of a lot of noise

Reply:
> I believe I would need `RefCell<S>` to be able to get a &mut S from &RefCell.

I see, and you're right. And owning those objects is precisely what you're trying to avoid.

Reply:
Kept thinking about this... I guess what I don't like (besides the multiplication of entities), is that this looks like a Rust "antipattern". I mean, in Rust the access mode (`mut` vs. `const`) is usually resolved by modifiers, or through a method call (`as_mut()` and such); not through object / struct definition.

Also, wouldn't make sense to just own `Storage`, or a copy of it? You can still have references for `Querier` and `Api`, but wouldn't make sense to pass `Storage` by value (or inside a `RefCell` if needed)?
I know this is against what you're trying to do here. Just pondering.

And, finally, maybe this can be resolved by passing a reference to a generic container, like `Deps`, but one that is not built on purpose, just to pass references, but is the original handler / container for those types. Not sure if this can be done given the current architecture / design decisions, though; and, if this wouldn't bring their own share of problems with multiple mutable access, borrowing, etc.